barcode_generator_product_variant/tests/test_barcode_generator_product_variant.py

```diff
@@ -0,0 +1,49 @@
+from odoo.exceptions import UserError
+from odoo.tests.common import SavepointCase
+
+
+class TestProductProductBarcodeWizard(SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        cls.barcode_rule = cls.env.ref(
+            "barcode_generator_product_variant.product_generated_barcode_demo"
+        )
+        cls.test_product = cls.env.ref("product.product_product_4_product_template")
+
+    def test_generate_missing_barcodes(self):
+        self.assertFalse(
+            any(i for i in self.test_product.mapped("product_variant_ids.barcode"))
+        )
+        self.assertTrue(self.test_product.has_missing_barcodes)
+
+        self.barcode_rule.is_default_rule = True
+        wizard = self.env["product.product.barcode.wizard"].create(
+            {
+                "product_tmpl_id": self.test_product.id,
+            }
+        )
+
+        # Barcode rule should be the default one
+        self.assertEqual(self.barcode_rule, wizard.barcode_rule_id)
+
+        wizard.action_generate_barcodes()
+        self.assertTrue(
+            any(i for i in self.test_product.mapped("product_variant_ids.barcode"))
```

Comment:
**todo**: shouldn't this be `all`?

Also same remark as above, generator is unnecessary

```suggestion
            all(self.test_product.mapped("product_variant_ids.barcode"))
```